### PR TITLE
feat: field compaction

### PR DESCRIPTION
**Pull request by arjansingh**
## Summary

When an LLM calls `github_list_issues`, the GitHub API returns ~100 fields per issue: nested user objects, avatar URLs, node IDs, permissions, template URLs. 

The model needs about 10 of those fields to decide what to do next. Multiply by 30 issues and you're burning 150KB of context on information the model will never use.

Field compaction fixes this. Integrations declare which fields matter for each tool. 

The server strips everything else automatically after every `execute` call. List and search responses are compact by default. 

When the LLM identifies a specific item from the compact list and calls a single-item get tool to inspect it, it still gets a full response back. That allows it to survey and drill-down.

## What Was Built

**Generic compaction engine** (`compact.go`): a JSON projector that supports three dot-notation patterns:
- `"title"` — keep scalar
- `"user.login"` — extract nested value
- `"labels[].name"` — pluck field from each array element

**Server integration**: opt-in `FieldCompactionIntegration` interface. Adapters implement `CompactSpec(toolName) ([]CompactField, bool)` to declare specs. Server applies compaction after `Execute`. Added a 50KB response cap as safety net to prevent super large responses from getting return.

**GitHub adapter**: This was the test case. Added ~50 list/search tool compaction specs. This is first adapter to opt in. Others can follow the same pattern.

## Benchmark Results

Tested it against the live MCP server, comparing compacted list responses vs raw get responses 

(N × single-item get = what the uncompacted list would cost).

| Workflow Question | Tool | Items | Compacted | Raw (N×get) | Ratio |
|---|---|---|---|---|---|
| "Which PRs need review?" | `list_prs` | 7 | 2.4 KB | ~108 KB | **~45x** |
| "Find the CI run that broke the build" | `list_workflow_runs` | 10 | 3.4 KB | ~89 KB | **~26x** |
| "Search for issues about X" | `search_issues` | 10 | 1.7 KB | ~31 KB | **~18x** |

All three workflow questions answered from a single compacted call — zero N+1 drill-downs needed.

## Design Decisions

**Token budget principle**: Field compaction should optimize for fewest total tokens across the entire task workflow, not smallest single response. A field like `requested_reviewers` costs ~50 bytes per PR in the list but saves a full 5KB drill-down call per PR. More fields per item = less compression per response, but fewer total tokens across the workflow.

**Opt-in, not mandatory**: Adapters that don't implement the interface are completely unaffected. No changes to `Integration` or `ToolResult`.

**Specs, not custom structs**: Instead of writing bespoke response types per tool, we can declare dot-notation field lists and let the field compaction engine handle it. We can even convert the slack adapter next to follow this pattern and remove the custom struct code.

## Test Plan

- [x] `make ci` passes (build, vet, test-race, lint, security)
- [x] Compaction engine: 31 unit tests covering all spec patterns, edge cases, idempotence
- [x] Server integration: mock tests for opt-in/skip/error/byte-cap paths
- [x] GitHub specs: parse validation, dispatch parity, no orphan specs, mutation exclusion
- [x] Search unwrap: handlers return arrays, compaction applies per-element

### CODE REVIEW

File: github/search.go [hunk 1]
```diff
@@ -9,11 +9,11 @@ import (
 
 func searchCode(ctx context.Context, g *integration, args map[string]any) (*mcp.ToolResult, error) {
 	opts := &gh.SearchOptions{ListOptions: listOpts(args)}
-	result, _, err := g.client.Search.Code(ctx, argStr(args, "query"), opts)
+	resp, _, err := g.client.Search.Code(ctx, argStr(args, "query"), opts)
 	if err != nil {
 		return errResult(err)
 	}
-	return jsonResult(result)
+	return jsonResult(resp.CodeResults)
```
[the remark this answers is not in the document]
Comment (arjansingh) [reply github/search.go]: Fixed. Search handlers now wrap the response back up with `total_count` before returning. Specs use `items[].field` to compact each element inside the wrapper.

This is what drove the nested `[]` engine support. `items[].labels[].name` needs two levels of array traversal.

File: server/server.go [hunk 2]
```diff
@@ -273,29 +278,72 @@ func (s *Server) handleExecute(ctx context.Context, req *mcpsdk.CallToolRequest)
...
+	if result.IsError {
+		return &mcpsdk.CallToolResult{
+			Content: []mcpsdk.Content{&mcpsdk.TextContent{Text: result.Data}},
+			IsError: true,
+		}, nil
+	}
+
+	result.Data = compactResult(integration, args.ToolName, result.Data)
+
+	if len(result.Data) > maxResponseBytes {
```
[the remark this answers is not in the document]
Comment (arjansingh) [reply server/server.go]: For now 50KB is working. Largest compacted response I've seen is ~3.4KB for 10 workflow runs. But you're right it could get tight as more adapters opt in.

I could make it a per integration setting on the `FieldCompactionIntegration` interface, so adapters with larger payloads can declare their own cap. Want me to go that direction?
